src/core/audio/AudioPortAudio.cpp

```diff
@@ -23,444 +23,231 @@
-	printf( "Output device: '%s' backend: '%s'\n", Pa_GetDeviceInfo( outDevIdx )->name, Pa_GetHostApiInfo( Pa_GetDeviceInfo( outDevIdx )->hostApi )->name );
-
-	// TODO: debug AudioEngine::pushInputFrames()
-	//m_supportsCapture = true;
-
-	_success_ful = true;
+	successful = true;
 }
 
-
-
-
 AudioPortAudio::~AudioPortAudio()
 {
 	stopProcessing();
-
-	if( !m_wasPAInitError )
-	{
-		Pa_Terminate();
-	}
-	delete[] m_outBuf;
 }
 
-
-
-
 void AudioPortAudio::startProcessing()
 {
-	m_stopped = false;
-	PaError err = Pa_StartStream( m_paStream );
-	
-	if( err != paNoError )
-	{
-		m_stopped = true;
-		printf( "PortAudio error: %s\n", Pa_GetErrorText( err ) );
-	}
+	PaError err = Pa_StartStream(m_paStream);
+	if (err != paNoError) { std::cerr << "Failed to start PortAudio stream: " << Pa_GetErrorText(err) << '\n'; }
 }
 
-
-
-
 void AudioPortAudio::stopProcessing()
 {
-	if( m_paStream && Pa_IsStreamActive( m_paStream ) )
+	if (m_paStream && Pa_IsStreamActive(m_paStream))
 	{
-		m_stopped = true;
-		PaError err = Pa_StopStream( m_paStream );
-	
-		if( err != paNoError )
-		{
-			printf( "PortAudio error: %s\n", Pa_GetErrorText( err ) );
-		}
+		PaError err = Pa_StopStream(m_paStream);
+		if (err != paNoError) { std::cerr << "Failed to stop PortAudio stream: " << Pa_GetErrorText(err) << '\n'; }
 	}
 }
 
-
-int AudioPortAudio::process_callback(const float* _inputBuffer, float* _outputBuffer, f_cnt_t _framesPerBuffer)
+int AudioPortAudio::processCallback(const float* inputBuffer, float* outputBuffer, f_cnt_t framesPerBuffer)
 {
-	if( supportsCapture() )
-	{
-		audioEngine()->pushInputFrames( (SampleFrame*)_inputBuffer, _framesPerBuffer );
-	}
-
-	if( m_stopped )
-	{
-		memset( _outputBuffer, 0, _framesPerBuffer *
-			channels() * sizeof(float) );
-		return paComplete;
-	}
+	std::fill_n(outputBuffer, framesPerBuffer * channels(), 0.0f);
 
-	while( _framesPerBuffer )
+	while (framesPerBuffer)
 	{
-		if( m_outBufPos == 0 )
+		if (m_outBufPos == 0)
 		{
-			// frames depend on the sample rate
-			const fpp_t frames = getNextBuffer( m_outBuf );
-			if( !frames )
-			{
-				m_stopped = true;
-				memset( _outputBuffer, 0, _framesPerBuffer *
-					channels() * sizeof(float) );
-				return paComplete;
-			}
-			m_outBufSize = frames;
+			const auto err = getNextBuffer(m_outBuf.data());
+			if (err == 0) { return paComplete; }
 		}
-		const auto min_len = std::min(_framesPerBuffer, m_outBufSize - m_outBufPos);
 
-		for( fpp_t frame = 0; frame < min_len; ++frame )
+		for (auto frame = std::size_t{0}; frame < framesPerBuffer; ++frame)
 		{
-			for( ch_cnt_t chnl = 0; chnl < channels(); ++chnl )
+			if (channels() == 1)
 			{
-				(_outputBuffer + frame * channels())[chnl] = AudioEngine::clip(m_outBuf[frame][chnl]);
+				outputBuffer[frame] = m_outBuf[frame].average();
+			}
+			else
+			{
+				outputBuffer[frame * channels()] = m_outBuf[m_outBufPos + frame][0];
+				outputBuffer[frame * channels() + 1] = m_outBuf[m_outBufPos + frame][1];
```

Comment:
Is the removal of `AudioEngine::clip` intentional?

Reply:
Not intentional, but it can be removed since by default PortAudio clips the audio IIRC. I'll probably leave it that way for performance and simplicity.